pkg/rancher/migrations.go

```diff
@@ -0,0 +1,79 @@
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	cattleNamespace        = "cattle-system"
+	upgradeMigrationConfig = "migrated"
+	forceLogoutKey = "forcedLogout"
+)
+
+// forceUpgradeLogout will delete all dashboard tokens forcing a logout.  This is useful when there is a major frontend
+//upgrade and we want all users to be sent to a central point.  This function will check for the `upgradeMigrationConfig`
+// configuration map and only run if the last migrated version is lower than the given `migrationVersion`.
+func forceUpgradeLogout(configMapController controllerv1.ConfigMapController, tokenController v3.TokenController, migrationVersion string) (err error) {
+	var upgradeMigrationConfigMap *v1.ConfigMap
+
+	upgradeMigrationConfigMap, err = configMapController.Get(cattleNamespace, upgradeMigrationConfig, metav1.GetOptions{})
+	if err != nil {
+		if k8serror.IsNotFound(err) { // if the config map does not exist we need to initialize it
+			upgradeMigrationConfigMap.Data = map[string]string{}
+		} else {
+			return
+		}
+	}
+
+	// clear out any revision ids or other metadata we don't want to save or initialize the cm for the first time
+	upgradeMigrationConfigMap = &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: upgradeMigrationConfig,
+			Namespace: cattleNamespace,
+		},
+		Data: upgradeMigrationConfigMap.Data,
+	}
+
+	// we only want to migrate if the previous installation was earlier than the migration version
+	if lastMigration, ok := upgradeMigrationConfigMap.Data[forceLogoutKey]; ok {
+		if lastMigration == "dev" || version.Compare(migrationVersion, lastMigration, "<=") {
+			return
+		}
+	}
+
+	logrus.Infof("detected %s upgrade, forcing logout for all web users", migrationVersion)
+
+	// list all tokens that were created for the dashboard
+	allTokens, err := tokenController.List(metav1.ListOptions{LabelSelector: labels.SelectorFromSet(labels.Set{tokens.TokenKindLabel: "session"}).String()})
+	if err != nil {
+		logrus.Debug("failed to list tokens for upgrade forced logout")
+		return
+	}
+
+	// log out all the dashboard users forcing them to be redirected to the login page
+	for i := 0; i < len(allTokens.Items); i++ {
+		err = tokenController.Delete(allTokens.Items[i].ObjectMeta.Name, &metav1.DeleteOptions{})
+		if err != nil && !k8serror.IsNotFound(err) {
+			logrus.Debug("failed to delete tokens for upgrade forced logout")
+			return
+		}
+	}
+
+	// record the migration being completed
+	upgradeMigrationConfigMap.Data[forceLogoutKey] = rancherversion.Version
+	_, err = configMapController.Create(upgradeMigrationConfigMap)
+	if k8serror.IsAlreadyExists(err) { // if someone else created the configmap we can just get it and check our value
+		return forceUpgradeLogout(configMapController, tokenController, migrationVersion)
```

Comment:
I don't think the recursive call is needed, if someone else already created it then the cleanup was already done so just exit

Reply:
This was a bit of forward thinking that we may want to use the migration configmap for other upgrade actions in the future.  I can rename it to `forcedLogout` if we want to make it single use instead.